### PR TITLE
Add `nonEmpty()` modifier for conditional defaults with `longestMatch()`

### LINKED ISSUE #79: Add `nonEmpty()` modifier to require parsers to consume at least one token
## Summary

Add a `nonEmpty()` modifier that requires a parser to consume at least one input token to succeed. This enables conditional default values in parsers using `longestMatch()`.

## Motivation

When building CLI parsers with conditional behavior, it's sometimes necessary to distinguish between "no options provided" and "at least one option provided". This is particularly useful when combining parsers with `longestMatch()` to implement patterns like:

1. `cli dev` (no options) → show help
2. `cli dev --key foo` (at least one option) → apply defaults and run
3. `cli dev --cwd ./path` (explicit value) → use provided value

This use case came up in https://github.com/dahlia/optique/discussions/78, where a user wanted to apply `withDefault()` conditionally based on whether any option was provided.

### Current behavior

`or()` already performs this check internally:

https://github.com/dahlia/optique/blob/f750e8e51b0697127e5cd704fe2ce18743759039/packages/core/src/constructs.ts#L1282

This means parsers that succeed without consuming input (like `constant()`) will cause `or()` to try the next alternative.

However, `longestMatch()` does not perform this check. It selects the parser that consumed the *most* tokens, even if that number is zero. When multiple parsers all consume zero tokens, `longestMatch()` selects the first one.

## Proposed API

```typescript
import { nonEmpty } from "@optique/core/modifiers";

const parser = longestMatch(
  nonEmpty(object({
    help: constant(false),
    cwd: withDefault(option("--cwd", string()), "./default-cwd"),
    key: optional(option("--key", string())),
  })),
  object({
    help: constant(true),
  })
);
```

**Behavior:**

- `cli dev` (no options):
  - First parser succeeds but consumes 0 tokens → rejected by `nonEmpty()`
  - Second parser succeeds and consumes 0 tokens → selected
  - Result: `{ help: true }`

- `cli dev --key foo` (at least one option):
  - First parser succeeds and consumes 2 tokens (`--key`, `foo`) → accepted by `nonEmpty()`
  - Second parser succeeds but consumes 0 tokens
  - Result: `{ help: false, cwd: "./default-cwd", key: "foo" }`

## Implementation

The `nonEmpty()` modifier wraps a parser and rejects its result if `consumed.length === 0`:

```typescript
export function nonEmpty<M extends Mode, T, TState>(
  parser: Parser<M, T, TState>,
): Parser<M, T, TState> {
  // Implementation similar to other modifiers
  // Check consumed.length after parsing
  // Return failure if consumed.length === 0, even if parser succeeded
}
```

## Use cases

- Conditional help display (show help only when no options provided)
- Conditional default values (apply defaults only when user provides some input)
- Disambiguation in `longestMatch()` when multiple branches can succeed with zero consumption

## Relationship to existing combinators

- `or()`: Already performs this check implicitly, so `nonEmpty()` is redundant inside `or()`
- `longestMatch()`: Does not perform this check, so `nonEmpty()` is useful here
- Other combinators (`merge()`, `concat()`): `nonEmpty()` can be used to enforce non-empty parsing in specific branches

## Alternative considered

Instead of adding `nonEmpty()`, we could:

1. Make `longestMatch()` reject zero-consumption parsers by default
2. Add a `requireConsumption` option to `longestMatch()`

These were rejected because:

- Changing `longestMatch()` behavior would be breaking
- `nonEmpty()` is more composable and can be used in other contexts
- It makes the requirement explicit at the parser level

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

- Add `nonEmpty()` modifier that requires a parser to consume at least one input token to succeed
- This enables conditional default values and help display logic when using `longestMatch()`

## Motivation

Closes #79

When building CLIs with optional flags that all have defaults, `longestMatch()` cannot distinguish between “no input provided” and “input with defaults applied” because both cases consume 0 tokens. The `nonEmpty()` modifier solves this by failing parsers that succeed without consuming any tokens.

Example use case:
```typescript
// cli dev         → Show help (no options)
// cli dev --key X → Run with options and defaults applied

const activeParser = nonEmpty(object({
  cwd: withDefault(option("--cwd", string()), "./default"),
  key: optional(option("--key", string())),
}));

const helpParser = object({ mode: constant("help") });
const parser = longestMatch(activeParser, helpParser);
```

## Changes

1. *Implementation* (*packages/core/src/modifiers.ts*)
   - New `nonEmpty()` function that wraps any parser
   - Supports both sync and async modes
   - Delegates `complete()`, `suggest()`, `getDocFragments()` to inner parser

2. *Tests* (*packages/core/src/modifiers.test.ts*)
   - Priority preservation
   - Success/failure scenarios based on token consumption
   - Integration with `longestMatch()`, `object()`, `withDefault()`, `multiple()`

3. *Documentation*
   - Concept guide section in *docs/concepts/modifiers.md*
   - Cookbook recipe in *docs/cookbook.md*
   - Example file at *examples/patterns/conditional-defaults.ts*

4. *Changelog* (*CHANGES.md*)

## Test plan

- [x] `mise test` passes on all runtimes (Deno, Node.js, Bun)
- [x] `mise check` passes (type checking, linting, formatting)
- [x] `cd docs && pnpm build` passes (Twoslash verification)
- [x] Example runs correctly: `deno run examples/patterns/conditional-defaults.ts`

### CODE REVIEW

File: packages/core/src/modifiers.ts [hunk 1]
```diff
@@ -1076,3 +1076,109 @@ export function multiple<M extends Mode, TValue, TState>(
...
+  return {
+    $mode: parser.$mode,
+    $valueType: parser.$valueType,
+    $stateType: parser.$stateType,
+    priority: parser.priority,
+    usage: parser.usage,
+    initialState: parser.initialState,
+    parse(context: ParserContext<TState>) {
+      if (isAsync) {
+        return parseAsync(context) as unknown as ParserResult<TState>;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts]: Addressed in 23270b79b20300f977ad3636c92e36c8483af6f9.

File: packages/core/src/modifiers.test.ts [hunk 2]
```diff
@@ -2547,3 +2548,230 @@ describe("state management edge cases", () => {
...
+      assert.equal(activeResult.value.mode, "active");
+      assert.equal(
+        (activeResult.value as { cwd: string }).cwd,
+        "./default",
+      );
+      assert.equal(
+        (activeResult.value as { key: string | undefined }).key,
+        "mykey",
+      );
+    }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.test.ts:2681]: Since this is test code where we've already asserted `activeResult.value.mode === "active"` on the previous line, the type assertion here is just to satisfy TypeScript's narrowing limitations with union types. The test will fail with a clear message if the mode is wrong, so the additional type guard would be redundant in this context.

File: packages/core/src/modifiers.test.ts [hunk 3]
```diff
@@ -2547,3 +2548,230 @@ describe("state management edge cases", () => {
...
+
+    // With --cwd option - activeParser wins
+    const cwdResult = parse(parser, ["--cwd", "./custom"]);
+    assert.ok(cwdResult.success);
+    if (cwdResult.success) {
+      assert.equal(cwdResult.value.mode, "active");
+      assert.equal((cwdResult.value as { cwd: string }).cwd, "./custom");
+    }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.test.ts:2689]: Same reasoning as above—this is test code where the assertion on line 2693 already verifies the mode. The type assertion is a pragmatic choice to work around TypeScript's union narrowing, and the test provides sufficient runtime verification.

File: docs/concepts/modifiers.md [hunk 4]
```diff
@@ -625,3 +625,137 @@ if (config.success) {
...
+~~~~ typescript twoslash
+import { object } from "@optique/core/constructs";
+import { multiple, nonEmpty, optional, withDefault } from "@optique/core/modifiers";
+import { parse } from "@optique/core/parser";
+import { argument, option } from "@optique/core/primitives";
+import { integer, string } from "@optique/core/valueparser";
+// ---cut-before---
+const configParser = nonEmpty(object({
+  // Required when any option is provided
+  files: multiple(argument(string({ metavar: "FILE" })), { min: 1 }),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/modifiers.md]: Addressed in 5cb20fb6be7bcb7825a959ceb8c0adf9ab370477.

File: packages/core/src/modifiers.test.ts [hunk 5]
```diff
@@ -2547,3 +2548,230 @@ describe("state management edge cases", () => {
...
+
+    const result = parse(parser, ["file1.txt", "file2.txt"]);
+    assert.ok(result.success);
+    if (result.success) {
+      assert.deepEqual(result.value, ["file1.txt", "file2.txt"]);
+    }
+
+    // No arguments - fails because nonEmpty requires consumption
+    const emptyResult = parse(parser, []);
+    assert.ok(!emptyResult.success);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.test.ts]: Addressed in 5cb20fb6be7bcb7825a959ceb8c0adf9ab370477.

File: packages/core/src/modifiers.ts [hunk 6]
```diff
@@ -1076,3 +1074,109 @@ export function multiple<M extends Mode, TValue, TState>(
...
+    // Check if inner parser consumed at least one token
+    if (result.consumed.length === 0) {
+      return {
+        success: false,
+        consumed: 0,
+        error: message`Parser must consume at least one token.`,
+      };
+    }
+    return result;
+  };
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts]: Addressed in 2aba523d6284dbcd829256023c8eb71b1c12cb1d.

File: packages/core/src/modifiers.test.ts [hunk 7]
```diff
@@ -2547,3 +2548,239 @@ describe("state management edge cases", () => {
...
+    if (activeResult.success) {
+      assert.equal(activeResult.value.mode, "active");
+      assert.equal(
+        (activeResult.value as { cwd: string }).cwd,
+        "./default",
+      );
+      assert.equal(
+        (activeResult.value as { key: string | undefined }).key,
+        "mykey",
+      );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.test.ts]: Addressed in 903c743a4e934b86e80b44766eee28e300539c3d.

File: packages/core/src/modifiers.test.ts [hunk 8]
```diff
@@ -2547,3 +2548,239 @@ describe("state management edge cases", () => {
...
+    }
+
+    // With --cwd option - activeParser wins
+    const cwdResult = parse(parser, ["--cwd", "./custom"]);
+    assert.ok(cwdResult.success);
+    if (cwdResult.success) {
+      assert.equal(cwdResult.value.mode, "active");
+      assert.equal((cwdResult.value as { cwd: string }).cwd, "./custom");
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.test.ts]: Addressed in 903c743a4e934b86e80b44766eee28e300539c3d.

File: packages/core/src/modifiers.ts [hunk 9]
```diff
@@ -1076,3 +1074,105 @@ export function multiple<M extends Mode, TValue, TState>(
...
+    }
+    // Check if inner parser consumed at least one token
+    if (result.consumed.length === 0) {
+      return {
+        success: false,
+        consumed: 0,
+        error: message`Parser must consume at least one token.`,
+      };
+    }
+    return result;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts:1136]: Both approaches are functionally equivalent, and I agree the combined condition more directly expresses the specific case `nonEmpty()` handles.

That said, I'll keep the current structure for now. The guard clause pattern separates concerns clearly (failure passthrough → zero-consumption check → success passthrough), and it allows TypeScript's type narrowing to work naturally after the `!result.success` check.
